File: backend/src/services/speech.service.ts
import { AssemblyAI } from 'assemblyai';
import { logger } from './logger.service';
import { createError } from '@/middleware/error.middleware';

const client = new AssemblyAI({
  apiKey: process.env.ASSEMBLYAI_API_KEY!
});

export class SpeechService {
  // Transcribe audio from file buffer
  async transcribeAudio(audioBuffer: Buffer, filename: string): Promise<string> {
    try {
      logger.info(`Starting transcription for file: ${filename}`);

      // Upload audio file
      const uploadUrl = await client.files.upload(audioBuffer);
      
      // Create transcription request
      const transcript = await client.transcripts.transcribe({
        audio_url: uploadUrl,
        language_code: 'en', // English language
        punctuate: true,
        format_text: true,
        speaker_labels: false, // Single speaker for interviews
        auto_highlights: true,
        sentiment_analysis: false,
        entity_detection: false,
        speech_model: 'best' // Use the best available model for highest accuracy
      });

      if (transcript.status === 'error') {
        throw new Error(transcript.error || 'Transcription failed');
      }

      logger.info(`Transcription completed successfully for: ${filename}`);
      return transcript.text || '';

    } catch (error) {
      logger.error('Error transcribing audio:', error);
      throw createError('Failed to transcribe audio', 500);
    }
  }

  // Transcribe audio from URL
  async transcribeFromUrl(audioUrl: string): Promise<string> {
    try {
      logger.info(`Starting transcription from URL: ${audioUrl}`);

      const transcript = await client.transcripts.transcribe({
        audio_url: audioUrl,
        language_code: 'en',
        punctuate: true,
        format_text: true,
        speaker_labels: false,
        auto_highlights: true,
        sentiment_analysis: false,
        entity_detection: false,
        speech_model: 'best'
      });

      if (transcript.status === 'error') {
        throw new Error(transcript.error || 'Transcription failed');
      }

      logger.info('Transcription completed successfully from URL');
      return transcript.text || '';

    } catch (error) {
      logger.error('Error transcribing audio from URL:', error);
      throw createError('Failed to transcribe audio', 500);
    }
  }

  // Check transcription status (for async operations)
  async getTranscriptionStatus(transcriptId: string): Promise<{
    status: string;
    text?: string;
    error?: string;
  }> {
    try {
      const transcript = await client.transcripts.get(transcriptId);
      
      return {
        status: transcript.status,
        text: transcript.text,
        error: transcript.error
      };

    } catch (error) {
      logger.error('Error checking transcription status:', error);
      throw createError('Failed to check transcription status', 500);
    }
  }

  // Start async transcription
  async startAsyncTranscription(audioBuffer: Buffer, filename: string): Promise<string> {
    try {
      logger.info(`Starting async transcription for file: ${filename}`);

      // Upload audio file
      const uploadUrl = await client.files.upload(audioBuffer);
      
      // Create transcription request
      const transcript = await client.transcripts.submit({
        audio_url: uploadUrl,
        language_code: 'en',
        punctuate: true,
        format_text: true,
        speaker_labels: false,
        auto_highlights: true,
        sentiment_analysis: false,
        entity_detection: false,
        speech_model: 'best'
      });

      logger.info(`Async transcription started with ID: ${transcript.id}`);
      return transcript.id;

    } catch (error) {
      logger.error('Error starting async transcription:', error);
      throw createError('Failed to start transcription', 500);
    }
  }

  // Get supported audio formats
  getSupportedFormats(): string[] {
    return [
      'mp3', 'mp4', 'm4a', 'wav', 'webm', 'ogg', 'flac', 
      'aac', '3gp', 'amr', 'wma', 'opus', 'aiff', 'au'
    ];
  }

  // Validate audio file
  validateAudioFile(filename: string, fileSize: number): { valid: boolean; error?: string } {
    const supportedFormats = this.getSupportedFormats();
    const fileExtension = filename.toLowerCase().split('.').pop();
    
    if (!fileExtension || !supportedFormats.includes(fileExtension)) {
      return {
        valid: false,
        error: `Unsupported audio format. Supported formats: ${supportedFormats.join(', ')}`
      };
    }

    // Maximum file size: 100MB
    const maxSize = 100 * 1024 * 1024;
    if (fileSize > maxSize) {
      return {
        valid: false,
        error: 'Audio file is too large. Maximum size: 100MB'
      };
    }

    return { valid: true };
  }
}

export const speechService = new SpeechService();

File: backend-fastapi/services/speech_service.py
import os
import time
import asyncio
from typing import Dict, Optional, Any
import assemblyai as aai
from dotenv import load_dotenv
from utils.logger import setup_logger
from utils.error_handler import async_retry_with_timeout, RetryableError, handle_exceptions

# Load environment variables
load_dotenv()

logger = setup_logger(__name__)

class SpeechService:
    """Service for speech-to-text operations using AssemblyAI"""
    
    def __init__(self):
        # Initialize AssemblyAI
        aai.settings.api_key = os.getenv("ASSEMBLYAI_API_KEY")
        if not aai.settings.api_key:
            raise ValueError("ASSEMBLYAI_API_KEY environment variable is required")
        
        self.transcriber = aai.Transcriber()
        logger.info("Speech Service initialized with AssemblyAI")

    @async_retry_with_timeout(max_retries=3, timeout=30)
    @handle_exceptions
    async def transcribe_audio(
        self, 
        audio_file_path: str,
        session_id: str,
        language_code: str = "en"
    ) -> Dict[str, Any]:
        """Transcribe audio file to text using AssemblyAI"""
        
        start_time = time.time()
        logger.info(f"Starting transcription for session: {session_id}")
        
        try:
            # Configure transcription settings
            config = aai.TranscriptionConfig(
                language_code=language_code,
                punctuate=True,
                format_text=True,
                speaker_labels=False,
                auto_highlights=False,
                sentiment_analysis=False,
                entity_detection=False
            )
            
            # Submit transcription job
            transcript = await asyncio.create_task(
                self._transcribe_async(audio_file_path, config)
            )
            
            processing_time = time.time() - start_time
            
            if transcript.status == aai.TranscriptStatus.error:
                error_msg = f"Transcription failed: {transcript.error}"
                logger.error(f"AssemblyAI error for session {session_id}: {error_msg}")
                raise RetryableError(error_msg)
            
            result = {
                "transcript": transcript.text or "",
                "confidence": transcript.confidence if hasattr(transcript, 'confidence') else None,
                "processing_time": processing_time,
                "language": language_code,
                "word_count": len(transcript.text.split()) if transcript.text else 0
            }
            
            logger.info(f"Transcription completed for session {session_id} in {processing_time:.2f}s")
            return result
            
        except Exception as e:
            processing_time = time.time() - start_time
            logger.error(f"Transcription failed for session {session_id}: {str(e)} (after {processing_time:.2f}s)")
            
            if "rate limit" in str(e).lower() or "quota" in str(e).lower():
                raise RetryableError(f"Rate limit or quota exceeded: {str(e)}")
            else:
                raise RetryableError(f"Transcription error: {str(e)}")

    async def _transcribe_async(self, audio_file_path: str, config: aai.TranscriptionConfig):
        """Async wrapper for AssemblyAI transcription"""
        
        # Run the synchronous transcription in a thread pool
        loop = asyncio.get_event_loop()
        transcript = await loop.run_in_executor(
            None,
            lambda: self.transcriber.transcribe(audio_file_path, config)
        )
        
        return transcript

    @handle_exceptions
    async def get_transcription_status(self, transcript_id: str) -> Dict[str, Any]:
        """Get the status of a transcription job"""
        
        try:
            transcript = await asyncio.create_task(
                self._get_transcript_async(transcript_id)
            )
            
            return {
                "id": transcript_id,
                "status": transcript.status.value,
                "text": transcript.text if transcript.status == aai.TranscriptStatus.completed else None,
                "confidence": transcript.confidence if hasattr(transcript, 'confidence') else None,
                "error": transcript.error if transcript.status == aai.TranscriptStatus.error else None
            }
            
        except Exception as e:
            logger.error(f"Failed to get transcription status for {transcript_id}: {str(e)}")
            raise RetryableError(f"Status check failed: {str(e)}")

    async def _get_transcript_async(self, transcript_id: str):
        """Async wrapper for getting transcript by ID"""
        
        loop = asyncio.get_event_loop()
        transcript = await loop.run_in_executor(
            None,
            lambda: aai.Transcript.get_by_id(transcript_id)
        )
        
        return transcript

    def validate_audio_file(self, file_path: str) -> bool:
        """Validate audio file format and size"""
        
        try:
            if not os.path.exists(file_path):
                logger.error(f"Audio file does not exist: {file_path}")
                return False
            
            file_size = os.path.getsize(file_path)
            max_size = 100 * 1024 * 1024  # 100MB limit
            
            if file_size > max_size:
                logger.error(f"Audio file too large: {file_size} bytes (max: {max_size})")
                return False
            
            if file_size == 0:
                logger.error(f"Audio file is empty: {file_path}")
                return False
            
            # Check file extension
            allowed_extensions = ['.wav', '.mp3', '.ogg', '.webm', '.m4a', '.flac']
            file_extension = os.path.splitext(file_path)[1].lower()
            
            if file_extension not in allowed_extensions:
                logger.error(f"Unsupported audio format: {file_extension}")
                return False
            
            return True
            
        except Exception as e:
            logger.error(f"Error validating audio file {file_path}: {str(e)}")
            return False

    def get_supported_formats(self) -> list:
        """Get list of supported audio formats"""
        
        return [
            {'extension': 'wav', 'mime_type': 'audio/wav', 'description': 'WAV audio'},
            {'extension': 'mp3', 'mime_type': 'audio/mpeg', 'description': 'MP3 audio'},
            {'extension': 'ogg', 'mime_type': 'audio/ogg', 'description': 'OGG audio'},
            {'extension': 'webm', 'mime_type': 'audio/webm', 'description': 'WebM audio'},
            {'extension': 'm4a', 'mime_type': 'audio/m4a', 'description': 'M4A audio'},
            {'extension': 'flac', 'mime_type': 'audio/flac', 'description': 'FLAC audio'}
        ]
